### PR TITLE
Rocket bug fix

### PR DESCRIPTION
**Pull request by simon33-2**
Fix bug which allows attacker to see result of first rocket attack before deciding subsequent rocket attacks' targets. This should not be possible. All targets should be decided before any results are resolved.

Required some two attempts to avoid picking up my other open PR

### CODE REVIEW

**Comment by simon33-2** [comment]
I know you don't like iterators. Care to suggest a better way of doing than to run through two iterable objects concurrently? I could create a class with two territories but it shouldn't be that hard.

Doesn't change old Rockets which only fired at a single territory.

**Comment by ron-murhammer** [comment]
@simon33-2 Could you instead use either a list of objects (the object contains the 2 objects) or use an ordered map? I'm definitely not a fan of iterating over 2 separate lists.

**Comment by simon33-2** [comment]
@ron-murhammer I agree but how do you do it without having to create a new class?

**Comment by ron-murhammer** [comment]
@simon33-2 LinkedHashMap if you don't want to create a new class.

**Comment by simon33-2** [comment]
@ron-murhammer that's done.

I'm having some trouble with gradle test on my computer but it is working on github so I guess it's ok.

File: src/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 1]
```diff
@@ -88,6 +90,7 @@ public void fireRockets(final IDelegateBridge bridge, final PlayerID player) {
   private void fireWW2V2(final IDelegateBridge bridge, final PlayerID player, final Set<Territory> rocketTerritories) {
     final GameData data = bridge.getData();
     final Set<Territory> attackedTerritories = new HashSet<>();
+    final LinkedHashMap<Territory,Territory> attackingTerritories = new LinkedHashMap<>();
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/RocketsFireHelper.java]: Left hand side should be the interface, not concrete type, ie: `HashMap< .. >`

File: src/games/strategy/triplea/delegate/RocketsFireHelper.java [hunk 2]
```diff
@@ -99,12 +102,13 @@ private void fireWW2V2(final IDelegateBridge bridge, final PlayerID player, fina
...
-        if (oneAttackPerTerritory) {
-          attackedTerritories.add(target);
-        }
-        fireRocket(player, target, bridge, territory);
+        attackedTerritories.add(target);
+        attackingTerritories.put(target,territory);
       }
     }
+    for( final Territory target : attackedTerritories ) {
+			fireRocket(player, target, bridge, attackingTerritories.get(target));
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/RocketsFireHelper.java]: formatting..

**Comment by simon33-2** [comment]
The user is prompted in getTarget.
The rocket is fired in "fireRocket". This is where the dice are rolled.

The problem comes about in the case where the attacker has two or more rockets (i.e. 1940 & most 1942 maps but not classic).

Consider this scenario. Attacker wants to disable an airbase or naval base. Fires a rocket at it. Has a second rocket which is in range. If the base is disabled by the first rocket, the second rocket can be targeted at something else but can target the base if it is still online. It is similar when a factory with a low production level is targeted.

While a lot of people play without tech, for those that do this bug is somewhat annoying because it affects game play.

There is a note (NA) for this in many maps. Shouldn't we be fixing these bugs?

I've noticed that in Revised (and some other versions) you are only permitted to target a territory once. I don't see this in the rule book. I think this is a bug in 1.809 and 1.9. Let's just fix this bug for now. Perhaps there is some reason?

**Comment by simon33-2** [comment]
Off topic, but I've noticed your comments in #18, where you lobbied to have the max column count reduced. Why would you want that? Better to have a long line than a continuation line IMO.

**Comment by panther2** [comment]
> I've noticed that in Revised (and some other versions) you are only permitted to target a territory once. I don't see this in the rule book. I think this is a bug in 1.809 and 1.9. 

The rule has been part of the official FAQ that had been published by Wizards/Avalon:

> Pgs. 20-21, Rocket Strikes: Delete the sentence "Choose an industrial complex within 3 spaces
and roll one die for each gun; however, each gun can inflict no more IPC loss than the territory's
income value." Replace it with "In each territory, one antiaircraft gun may attack an industrial
complex within 3 spaces, **though each industrial complex may be attacked by only one rocket
launcher in a turn**. The maximum loss each rocket can inflict is the target territory's income value."

**Comment by DanVanAtta** [comment]
Suggested some updates here: https://github.com/simon33-2/triplea/pull/1 
With those applied I do not have any objections to this update.  I've tested out the v1, v3 and WWII global behavior -  seems to do what has been promised.

**Comment by DanVanAtta** [comment]
Re: max column

In short, there is inherit complexity in long lines. Longer than 50-60 is bad, 80-100 is being very generous, 140 IMO is too much.

Google did some research into code formatting effects on productivity and found horizontal  scrolling of code  reduced productivity. 

Consider a web page for example, scrolling left and right to finish reading the line is very annoying, and a tiny bit time consuming. Add this up over hundreds of lines, and over long periods of time (this software already is a decade old, and still going), and the cost adds up.

To another point, humans are only able to understand / comprehend so much. It's for that reason despite screen resolutions being on average better, web page text still is not very long and tends to be in relatively consumable paragraph widths. That is roughly apparent even in how github formats the messages in issues, 106 characters is roughly the limit.

Humans can consume about 50-60 characters per line before they start becoming overwhelmed.IE: there is only so much info you can consume at once, and when something is so chained together on one line out to 140+ characters, it's a lot to take in and makes the code hard to understand. 

The general recommendation for 80-100 that google recommends is pretty generous. 

A final factor is it does create motivation to write code in pretty concise and 'small' chunks. If you are typing out to the 100th column, chances are things should be broken up anyways.

**Comment by simon33-2** [comment]
Ok, I've merged in those changes @DanVanAtta 

Don't know that I see it the same way on the line length. What we see in this app is lots and lots of continuation lines which are far worse IMO. I find that I merge them together in some cases so I can read it more easily.

**Comment by DanVanAtta** [comment]
The 60 vs 80 vs 140 is not really about line wrapping necessarily. You could write one statement that is 1000 characters long and wrap it at arbitrary character widths easily. Really the point is the abstractions you create and use should be concise enough that statements can fit entirely on 60 to max 80 characters.

**Comment by simon33-2** [comment]
Thanks for merging this one @DanVanAtta

**Pull request merged**